src/plots/linePlot.ts

```diff
@@ -425,6 +449,72 @@ export module Plots {
       return filteredDataIndices;
     }
 
+    private _filterDownsampling(dataset: Dataset, indices: number[]) {
+      let data = dataset.data();
+      let filteredIndices: number[] = [];
+      let minScaledValue: number;
+      let maxScaledValue: number;
+      let currentSlope: number;
+      let scaledXAccessor = Plot._scaledAccessor(this.x());
+      let scaledYAccessor = Plot._scaledAccessor(this.y());
+
+      if (indices.length === 0) {
+        return filteredIndices;
+      }
+      filteredIndices.push(indices[0]);
+
+      let belongingToCurrentBucket = (i: number) => {
+        let p1x = scaledXAccessor(data[indices[i]], indices[i], dataset);
+        let p1y = scaledYAccessor(data[indices[i]], indices[i], dataset);
+        let p2x = scaledXAccessor(data[indices[i + 1]], indices[i + 1], dataset);
+        let p2y = scaledYAccessor(data[indices[i + 1]], indices[i + 1], dataset);
+        if (currentSlope == null) {
+          currentSlope = (Math.floor(p1x) === Math.floor(p2x)) ? Infinity : (p2y - p1y) / (p2x - p1x);
+          minScaledValue = (currentSlope === Infinity) ? p1y : p1x;
+          maxScaledValue = minScaledValue;
+          return true;
+        }
+        if (currentSlope === Infinity) {
+          return Math.floor(p1x) === Math.floor(p2x);
+        } else {
+          let expectedP2y = p1y + (p2x - p1x) * currentSlope;
+          return Math.floor(p2y) === Math.floor(expectedP2y);
+        }
+      };
+
+      for (let i = 0; i < indices.length - 1; ) {
+        currentSlope = null;
```

Comment:
The fact that we have to reset `currentSlope` every time signals the fact that this variable should only be scoped inside the for loop.

If the item prventing this is that `currentSlope` is defined inside the function, then let's use `currentSlope` as an argument to a function rather than having it be a more globalized variable.